File: intern/42.trapping-rain-water.js
/*
 * @lc app=leetcode id=42 lang=javascript
 *
 * [42] Trapping Rain Water
 */

// @lc code=start
/**
 * TLE
 * @param {number[]} height
 * @return {number}
 */
var trap = function (height) {
  const LENGTH = height.length;
  if (LENGTH < 3) {
    return 0;
  }

  let water = 0;

  let left = 0;
  let right = LENGTH - 1;
  let maxLeft = 0;
  let maxRight = LENGTH - 1;

  while (left <= right) {
    if (height[left] <= height[right]) {
      if (height[left] > height[maxLeft]) {
        maxLeft = left;
      } else {
        water += height[maxLeft] - height[left];
      }
      left++;
    } else {
      if (height[right] > height[maxRight]) {
        maxRight = right;
      } else {
        water += height[maxRight] - height[right];
      }
      right--;
    }
  }

  return water;
};
// @lc code=end

File: prev-2024/42.trapping-rain-water.cpp
/*
 * @lc app=leetcode id=42 lang=cpp
 *
 * [42] Trapping Rain Water
 */
#include <cmath>
#include <iostream>
#include <vector>

using std::vector;

// @lc code=start
class Solution {
 public:
  int trap(const vector<int>& heights) {
    const int N = heights.size();
    int left = 0;
    int right = N - 1;
    int leftMaxHeight = 0;
    int rightMaxHeight = 0;
    int water = 0;

    while (left <= right) {
      if (heights[left] < heights[right]) {
        if (leftMaxHeight < heights[left]) {
          leftMaxHeight = heights[left];
        } else {
          water += leftMaxHeight - heights[left];
        }
        left++;
      } else {
        if (rightMaxHeight < heights[right]) {
          rightMaxHeight = heights[right];
        } else {
          water += rightMaxHeight - heights[right];
        }
        right--;
      }
    }

    return water;
  }
};
// @lc code=end